### PR TITLE
Improve configuration file generation

### PR DESCRIPTION
**Pull request by AnzeXie**
Added options for configuration generation and config arguments to preprocess.py. 
E.g.
--generate_config, --config_dir, --model.embedding_size

Added tests for above features.

### CODE REVIEW

File: tools/preprocess.py [hunk 1]
```diff
@@ -496,9 +478,28 @@ def extract_file(filepath):
     }
 
     if dataset_dict.get(args.dataset) != None:
-        dataset_dict.get(args.dataset)(args.output_directory, args.num_partitions)
-        output_bash_cmds(args.output_directory, args.dataset)
```
Comment (JasonMoho) [inline tools/preprocess.py:500]: output_bash_cmds is no longer used but still imported in this file.

**Review by JasonMoho — CHANGES_REQUESTED** [review]

File: tools/preprocess.py [hunk 2]
```diff
@@ -453,20 +466,60 @@ def extract_file(filepath):
     return filepath.parent
 
 
-if __name__ == "__main__":
-    parser = argparse.ArgumentParser(description='Preprocess Datasets')
-    parser.add_argument('dataset', metavar='dataset', type=str, help='Dataset to preprocess')
-    parser.add_argument('output_directory', metavar='output_directory', type=str, help='Directory to put graph data')
-    parser.add_argument('--num_partitions', metavar='num_partitions', required=False, type=int, default=1,
+def setArgs():
+    parser = argparse.ArgumentParser(
```
Comment (JasonMoho) [inline tools/preprocess.py:470]: Usage is still unclear, we should improve this according to our conversation over slack

**Review by JasonMoho — CHANGES_REQUESTED** [review]

File: tools/preprocess.py [hunk 3]
```diff
@@ -453,20 +392,63 @@ def extract_file(filepath):
     return filepath.parent
 
 
-if __name__ == "__main__":
+def setArgs():
     parser = argparse.ArgumentParser(description='Preprocess Datasets')
     parser.add_argument('dataset', metavar='dataset', type=str, help='Dataset to preprocess')
     parser.add_argument('output_directory', metavar='output_directory', type=str, help='Directory to put graph data')
```
[the remark this answers is not in the document]
Comment (AnzeXie) [reply tools/preprocess.py]: preprocessed data and configuration file generated are put into the same directory now (specified by args.output_directory)

File: tools/config_templates/default_configs.txt [hunk 4]
```diff
@@ -0,0 +1,66 @@
+general.device=CPU*CPU*GPU
```
[the remark this answers is not in the document]
Comment (AnzeXie) [reply tools/config_templates/default_configs.txt:1]: The first value before "*" is the default value. Values after first * are the valid values for the option.

File: tools/preprocess.py [hunk 5]
```diff
@@ -453,20 +466,60 @@ def extract_file(filepath):
...
+                       '[--<section>.<option> [value]] ' +
+                        '\ndataset output_directory')), add_help=False)
+    parser.add_argument('dataset', metavar='dataset',
+                        type=str, help='Dataset to preprocess')
+    parser.add_argument('output_directory', metavar='output_directory',
+                        type=str, help='Directory to put graph data')
+    parser.add_argument('--num_partitions', metavar='num_partitions',
+                        required=False, type=int, default=1,
                         help='Number of partitions to split the edges into')
+    parser.add_argument('--generate_config', '-gc', metavar='generate_config',
```
[the remark this answers is not in the document]
Comment (AnzeXie) [reply tools/preprocess.py:482]: Added the multi-gpu config generation option

**Review by JasonMoho — DISMISSED** [review]
Ship it! 🚢

**Review by JasonMoho — APPROVED** [review]

**Pull request merged**